### PR TITLE
blueprintwizard: remove empty customizations (RHEL-28534)

### PR DESCRIPTION
**Pull request by kingsleyzissou**
This commit sanitizes the form data and makes sure that the values of
the form fields are truly empty. For example, the filesystem customization
option was being initialized to an empty array. This was causing `edge-commit`
and `edge-commit` image types to break since they are not allowed
filesystem customizations.

Additionally, this commit ensures that all the fields are undefined if they have
not been set. This ensures that the blueprint is saved correctly and the below
no longer happens:

```
[customizations]
[customizations.kernel]
append = ""
[customizations.timezone]
[customizations.locale]
[customizations.firewall]
[customizations.firewall.services]
[customizations.services]
[customizations.fdo]
[customizations.ignition]
[customizations.ignition.firstboot]
```

### CODE REVIEW

File: src/components/Wizard/BlueprintWizard.js [hunk 1]
```diff
@@ -87,6 +87,20 @@ const BlueprintWizard = (props) => {
         ? formValues
         : formApi.getState().values;
     const blueprintData = formStateToBlueprint(formState);
+
+    // when customizations are provided, the users and filesystem customizations
+    // are initialized to an empty array. This creates issues for image types that
+    // aren't allowed these customizations
+    if (blueprintData.customizations) {
```
Comment (kingsleyzissou) [inline src/components/Wizard/BlueprintWizard.js:94]: Ideally we would make sure that this isn't set in the first place. But the data driven forms state is a bit awkward

File: src/components/Wizard/BlueprintWizard.js [hunk 2]
```diff
@@ -87,6 +87,20 @@ const BlueprintWizard = (props) => {
...
+      if (blueprintData.customizations.filesystem?.length === 0) {
+        delete blueprintData.customizations.filesystem;
+      }
+
+      if (blueprintData.customizations.user?.length === 0) {
```
Comment (kingsleyzissou) [inline src/components/Wizard/BlueprintWizard.js:99]: We may actually need to check some other fields such as services, locale etc if the user has stepped through those form steps, some of those values might be empty too.

Comment (kingsleyzissou) [reply src/components/Wizard/BlueprintWizard.js:99]: Hmm I've checked the other steps and we get this when I click through each step but don't edit anything:

```
[customizations]
[customizations.kernel]
append = ""
[customizations.timezone]
[customizations.locale]
[customizations.firewall]
[customizations.firewall.services]
[customizations.services]
[customizations.fdo]
[customizations.ignition]
[customizations.ignition.firstboot]
```

However, it seems like the only one that was causing issues was the filesystem customizations.

**Comment by kingsleyzissou** [comment]
This PR fixes both:
- RHEL-4668
- RHEL-28534

But RHEL-28534 is longer standing and was initially a blocker to the RHEL-9.4 release

File: src/helpers.js [hunk 3]
```diff
@@ -111,7 +155,15 @@ const formStateToCustomizations = (customizations) => {
     services,
   };
 
-  return customizationsParsed;
+  // if a user steps through each component in the form without filling any
+  // of the steps out, there is a possibility that each form field is empty.
+  // if this is the case, we want to return an empty (or undefined) `customizationsParsed`
+  // object
+  const everythingUndefined = Object.keys(customizationsParsed).every(
```
Comment (kingsleyzissou) [inline src/helpers.js:162]: I could simplify this by only initialising the `customizationsParsed` if a field is not empty... but this was a quick win

**Review by regexowl — APPROVED** [review]
Haven't tested this locally, but the changes make sense and the tests are passing :sweat_smile:

**Pull request merged**